### PR TITLE
Fix FoldableBlock.joinIntersecting(), fix newline handling after a service comment, export TextEditingValue.typed, add TextEditingValue.select

### LINKED ISSUE #136: CodeController.foldImports does not fold a multiline import in Go
Reproducible with MinimalWordCount example in Playground. In short:

```go
// license
// license
package main

import (
  "context"
  "fmt"
  "regexp"

  "github.com/apache/beam/sdks/v2/go/pkg/beam"
  "github.com/apache/beam/sdks/v2/go/pkg/beam/io/textio"
  "github.com/apache/beam/sdks/v2/go/pkg/beam/runners/direct"
  "github.com/apache/beam/sdks/v2/go/pkg/beam/transforms/stats"

  _ "github.com/apache/beam/sdks/v2/go/pkg/beam/io/filesystem/gcs"
  _ "github.com/apache/beam/sdks/v2/go/pkg/beam/io/filesystem/local"
)

```

### PR DESCRIPTION
**Pull request by alexeyinkin**
- Resolves #136

### CODE REVIEW

File: test/src/code_field/text_editing_value_test.dart [hunk 1]
```diff
@@ -6,76 +6,76 @@ import '../common/lorem_ipsum.dart';
 
 void main() {
   group('TextEditingValue ', () {
-    test('wordAtCursorStart', () {
+    test('wordAtCursor', () {
```
Comment (alexeyinkin) [inline test/src/code_field/text_editing_value_test.dart:9]: Just changed the order of tests here, and it recognized them as changed lines.

File: lib/src/folding/foldable_block.dart [hunk 2]
```diff
@@ -55,38 +55,102 @@ class FoldableBlock extends InclusiveRange with EquatableMixin {
...
+      // Throw the block into the current nesting hierarchy...
+      ancestors.add(bubble);
+      ancestorIndexToOverallIndex.add(overallIndex);
+
+      int bubbleIndexInAncestors = ancestors.length - 1;
+      int bubbleOverallIndex = overallIndex;
+
+      // And fix every violation of the hierarchy by bubbling the block up,
+      // removing non-ancestors from the working list, and joining when needed.
+      for (int ancestorIndex = ancestors.length - 1; --ancestorIndex >= 0;) {
```
Comment (yescorp) [inline lib/src/folding/foldable_block.dart:114]: I think this way it is more readable.
```dart
for (int ancestorIndex = ancestors.length - 2; ancestorIndex >= 0; --ancestorIndex)
```
Also emphasizes on why the first iteration should not consider last element of the list.

File: lib/src/folding/foldable_block.dart [hunk 3]
```diff
@@ -55,38 +55,102 @@ class FoldableBlock extends InclusiveRange with EquatableMixin {
 
     return false;
   }
+
+  bool isSameLines(FoldableBlock other) {
```
Comment (Malarg) [inline lib/src/folding/foldable_block.dart:59]: `areSameLines`?

File: lib/src/folding/foldable_block.dart [hunk 4]
```diff
@@ -55,38 +55,102 @@ class FoldableBlock extends InclusiveRange with EquatableMixin {
...
+    if (a.type == FoldableBlockType.imports) {
+      return FoldableBlockType.imports;
+    }
+
+    if (b.type == FoldableBlockType.imports) {
+      return FoldableBlockType.imports;
+    }
+
+    return FoldableBlockType.union;
+  }
```
Comment (Malarg) [inline lib/src/folding/foldable_block.dart:81]:
```suggestion
  static FoldableBlockType _getJoinType(FoldableBlock a, FoldableBlock b) {
    final containsImports = a.isImports || b.isImports;
    return containsImports ? FoldableBlockType.imports : FoldableBlockType.union;
  }
```

File: lib/src/code_field/text_editing_value.dart [hunk 5]
```diff
@@ -152,4 +152,39 @@ extension TextEditingValueExtension on TextEditingValue {
...
+    );
+  }
+
+  TextEditingValue? select(Pattern pattern, [int start = 0]) {
+    if (pattern == '') {
+      throw AssertionError('Cannot search for an empty pattern');
+    }
+
+    final position = text.indexOf(pattern, start);
+
```
Comment (nausharipov) [inline lib/src/code_field/text_editing_value.dart:170]: Is there a rule for spacing? I expect no newline between related `position` initialization and and the `if` statement.

File: lib/src/folding/foldable_block.dart (hunk 4 above)
Comment (alexeyinkin) [reply lib/src/folding/foldable_block.dart:81]: Now imagine that we need to preserve the type if the blocks are of the same type. With this code, we would have to change lines, and also `?:` must be removed.

Those cut-offs are deliberate. They allow a lot of extension by just adding lines.

File: lib/src/folding/foldable_block.dart (hunk 3 above)
Comment (alexeyinkin) [reply lib/src/folding/foldable_block.dart:59]: `is` refers to the object and not its lines: `block.is...`

File: lib/src/folding/foldable_block.dart (hunk 2 above)
Comment (alexeyinkin) [reply lib/src/folding/foldable_block.dart]: `+`

File: lib/src/folding/foldable_block.dart [hunk 6]
```diff
@@ -55,38 +55,104 @@ class FoldableBlock extends InclusiveRange with EquatableMixin {
...
-        );
-        removeAt(i);
-        i = max(i - 2, 0);
+    // A working list to lay down blocks to compare.
+    // As we iterate blocks, it will contain the current nesting hierarchy.
+    final ancestors = <FoldableBlock>[];
+    final ancestorIndexToOverallIndex = <int>[];
+
+    for (int overallIndex = 0; overallIndex < length; overallIndex++) {
+      FoldableBlock bubble = this[overallIndex];
```
Comment (nausharipov) [inline lib/src/folding/foldable_block.dart:103]: A description for why it's named `bubble` might be useful.

File: lib/src/folding/foldable_block.dart [hunk 7]
```diff
@@ -55,38 +55,104 @@ class FoldableBlock extends InclusiveRange with EquatableMixin {
...
   void sortByStartLine() {
     sort((a, b) => a.firstLine - b.firstLine);
   }
 
-  /// Joins intersecting blocks in list.
-  /// Expected that list is sorted by start line.
+  /// Joins intersecting blocks.
+  ///
+  /// This list must be sorted by firstLine.
   void joinIntersecting() {
```
Comment (nausharipov) [inline lib/src/folding/foldable_block.dart:92]: Would helper methods improve readability?

File: test/src/code_field/text_editing_value_select_test.dart [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+
+import 'package:flutter/widgets.dart';
+import 'package:flutter_code_editor/flutter_code_editor.dart';
+import 'package:flutter_test/flutter_test.dart';
+
+void main() {
+  group('TextEditingValue ', () {
+    test('select', () {
+      final examples = [
+        //
```
Comment (nausharipov) [inline test/src/code_field/text_editing_value_select_test.dart:12]: Is this empty comment deliberate?

File: lib/src/folding/foldable_block.dart (hunk 6 above)
Comment (alexeyinkin) [reply lib/src/folding/foldable_block.dart:103]: `+`

File: test/src/code_field/text_editing_value_select_test.dart (hunk 8 above)
Comment (alexeyinkin) [reply test/src/code_field/text_editing_value_select_test.dart:12]: Yes, it preserves empty lines between examples.

**Pull request merged**